data/data_generator.jl

```diff
@@ -0,0 +1,582 @@
+comb1st_indices = Dict{UInt32,Int}()
+comb1st_indices_sorted_keys = Origin(0)(UInt32[])
+comb2nd_indices = Dict{UInt32,Int}()
+comb2nd_indices_sorted_keys = Origin(0)(UInt32[])
+comb2nd_indices_nonbasic = Set{UInt32}()
+comb_array = Origin(0)(Vector{Dict{Int,UInt32}}())
+for char in char_props
+    if isnothing(char.decomp_type) && !isnothing(char.decomp_mapping) &&
+            length(char.decomp_mapping) == 2 && !isnothing(char_hash[char.decomp_mapping[1]]) &&
+            char_hash[char.decomp_mapping[1]].combining_class == 0 &&
+            char.code ∉ exclusions
+        dm0 = char.decomp_mapping[1]
+        dm1 = char.decomp_mapping[2]
+        if !haskey(comb1st_indices, dm0)
+            comb1st_indices[dm0] = length(comb1st_indices)
+            push!(comb1st_indices_sorted_keys, dm0)
+            push!(comb_array, Dict{Int,UInt32}())
+            @assert length(comb1st_indices) == length(comb_array)
+        end
+        if !haskey(comb2nd_indices, dm1)
+            push!(comb2nd_indices_sorted_keys, dm1)
+            comb2nd_indices[dm1] = length(comb2nd_indices)
+        end
+        @assert !haskey(comb_array[comb1st_indices[dm0]], comb2nd_indices[dm1])
+        comb_array[comb1st_indices[dm0]][comb2nd_indices[dm1]] = char.code
+        if char.code > 0xFFFF
+            push!(comb2nd_indices_nonbasic, dm1)
+        end
+    end
+end
+
+comb_indices = Dict{UInt32,Int}()
+comb1st_indices_lastoffsets = Origin(0)(zeros(Int, length(comb1st_indices)))
+comb1st_indices_firstoffsets = Origin(0)(zeros(Int, length(comb1st_indices)))
+let
+    cumoffset = 0
+    for dm0 in comb1st_indices_sorted_keys
+        index = comb1st_indices[dm0]
+        first = nothing
+        last = nothing
+        offset = 0
+        for b in eachindex(comb2nd_indices_sorted_keys)
+            dm1 = comb2nd_indices_sorted_keys[b]
+            if haskey(comb_array[index], b)
+                if isnothing(first)
+                    first = offset
+                end
+                last = offset
+                if dm1 in comb2nd_indices_nonbasic
+                    last += 1
+                end
+            end
+            offset += 1
+            if dm1 in comb2nd_indices_nonbasic
+                offset += 1 
+            end
+        end
+        comb1st_indices_firstoffsets[index] = first
+        comb1st_indices_lastoffsets[index] = last
+        @assert !haskey(comb_indices, dm0)
+        comb_indices[dm0] = cumoffset
+        cumoffset += last - first + 1 + 2
+    end
+
+    offset = 0
+    for dm1 in comb2nd_indices_sorted_keys
+        @assert !haskey(comb_indices, dm1)
+        comb_indices[dm1] = 0x8000 | (comb2nd_indices[dm1] + offset)
+        @assert comb2nd_indices[dm1] + offset <= 0x4000
+        if dm1 in comb2nd_indices_nonbasic
+            comb_indices[dm1] |= 0x4000
+            offset += 1
+        end
+    end
+end
+
+function utf16_encode(utf32_seq)
```

Comment:
Can’t Julia’s `transcode` function do this?

Reply:
It seems like it can't, not directly.  But it seems we can use `transcode` if we go via UTF-8. I could do that if you prefer?

```julia
julia> utf32_seq = transcode(UInt32, "a🎏") # for example
2-element Vector{UInt32}:
 0x00000061
 0x0001f38f

julia> transcode(UInt16, utf32_seq)
ERROR: MethodError: no method matching transcode(::Type{UInt16}, ::Vector{UInt32})

Closest candidates are:
  transcode(::Type{UInt8}, ::Vector{<:Union{Int32, UInt32}})
   @ Base strings/cstring.jl:170
  transcode(::Type{UInt16}, ::AbstractVector{UInt8})
   @ Base strings/cstring.jl:181
  transcode(::Type{String}, ::Any)
   @ Base strings/cstring.jl:179
  ...

Stacktrace:
 [1] top-level scope
   @ REPL[91]:1

julia> transcode(UInt16, transcode(String, utf32_seq))
3-element Vector{UInt16}:
 0x0061
 0xd83c
 0xdf8f
```

Reply:
Yes, let's just use `transcode` here — even though it is slightly less efficient to go through utf-8, maximizing efficiency isn't a big concern.

Reply:
Efficiency is no concern but I do prefer the explicit version because it mirrors the decoding on the C side in a much more obvious way.

Reply:
It's a lot clearer not to duplicate functionality from Base here.